File: parser/Node.cpp

#pragma once

#include <vector>
#include <string>
#include <variant>

#include "Type.cpp"
#include "utils.cpp"

class Node
{
 private:
	Type type;
	std::vector<std::variant<Node, std::nullptr_t>> children;

 public:
	Node();
	Node(Type type);
	~Node();
	auto prependChild(Node child);
	auto appendChild(Node child);
	auto appendChildren(std::vector<std::variant<Node, std::nullptr_t>> children);
	std::string toJson() const;

	std::variant<Node, std::nullptr_t> getFirstByType(Type) const;
	bool hasChildOfType(Type) const;
	std::vector<std::variant<Node, std::nullptr_t>> getChildren() const;
	int hasChildren() const;
	Type getType() const;
	auto setType(Type);

	auto deleteChild(Node&);
};

// -----------------------------------------------------------------------------

Node::Node()
{
	std::cout << "NONONONO!!!!\n";
}

auto Node::setType(Type type)
{
	this->type = type;
	return *this;
}

auto Node::deleteChild(Node& child_to_remove)
{
	for (auto& child : children) {
		if (&std::get<Node>(child) == &child_to_remove) {
			child = nullptr;
		}
	}

	return *this;
}

Node::Node(Type type)
{
	//std::cout << "(node " << type << ")";
	setType(type);
}

Node::~Node()
{
	//std::cout << "(~node " << type << ")";
}

auto Node::prependChild(Node child)
{
	children.insert(children.begin(), child);
	return *this;
}

auto Node::appendChild(Node child)
{
	if (type == Type::UNION) {
		std::cout << "Appending to UNION (" << this->toJson() << "): " << child.toJson() << "\n";
	}
	children.push_back(child);
	if (type == Type::UNION) {
		std::cout << "After Appending to UNION (" << this->toJson() << ")\n\n";
	}
	return *this;
}

auto Node::appendChildren(std::vector<std::variant<Node, std::nullptr_t>> children)
{
	for (auto child : children) {
		if (std::holds_alternative<Node>(child)) {
			this->appendChild(std::get<Node>(child));
		}
	}
	return *this;
}

std::string Node::toJson() const
{
	std::string result;
	int count = children.size();
	bool human_readable = true;
	int i = 1;

	if (human_readable) {
		result += nodeTypeToString(type);

		if (count) { result += '<'; }
		for (auto child : children) {
			result += std::holds_alternative<Node>(child) ? std::get<Node>(child).toJson() : "~";
			if (i < count) { result += ", "; }
			i++;
		}
		if (count) { result += '>'; }
	} else {
		result += "{\"type\":\"" + nodeTypeToString(type) + "\"";

		if (count) { result += ",\"children\":["; }
		for (auto child : children) {
			result += std::holds_alternative<Node>(child) ? std::get<Node>(child).toJson() : "~";
			if (i < count) { result += ','; }
			i++;
		}
		if (count) { result += ']'; }

		result += '}';
	}

	return result;
}

std::variant<Node, std::nullptr_t> Node::getFirstByType(Type type) const
{
	if (this->type == type) {
		return *this;
	}

	for (auto child : children) {
		if (std::get<Node>(child).getType() == type) {
			return child;
		}

		if (std::get<Node>(child).hasChildren()) {
			std::variant<Node, std::nullptr_t> result = std::get<Node>(child).getFirstByType(type);
			if (std::holds_alternative<Node>(result)) {
				return std::get<Node>(result);
			}
		}
	}

	return nullptr;
}

bool Node::hasChildOfType(Type type) const
{
	for (auto child : children) {
		if (std::get<Node>(child).getType() == type) {
			return true;
		}
	}

	return false;
}

std::vector<std::variant<Node, std::nullptr_t>> Node::getChildren() const
{
	return children;
}

int Node::hasChildren() const
{
	return children.size();
}

Type Node::getType() const
{
	return type;
}

File: src/Parser/Node.php
<?php

namespace Zheltikov\TypeAssert\Parser;

class Node implements \JsonSerializable
{
    /**
     * @var Type
     */
    protected $type;

    /**
     * @var \Zheltikov\TypeAssert\Parser\Node[]
     */
    protected $children;

    public function __construct(Type $type)
    {
        $this->type = $type;
        $this->children = [];
    }

    /**
     * @param \Zheltikov\TypeAssert\Parser\Node|null $child
     * @return $this
     */
    public function appendChild(?Node $child = null): self
    {
        if ($child !== null) {
            $this->children[] = $child;
        }

        return $this;
    }

    /**
     * @param iterable|null $children
     * @return $this
     */
    public function appendChildren(?iterable $children = null): self
    {
        if ($children !== null) {
            foreach ($children as $child) {
                $this->appendChild($child);
            }
        }

        return $this;
    }

    /**
     * @return \Zheltikov\TypeAssert\Parser\Node[]
     */
    public function &getChildren(): array
    {
        return $this->children;
    }

    /**
     * @return array
     */
    public function jsonSerialize(): array
    {
        return [
            'type' => $this->type->getKey(),
            'children' => $this->children,
        ];
        /*return [
            $this->type->getKey(),
            ...$this->children,
        ];*/
    }

    /**
     * @param \Zheltikov\TypeAssert\Parser\Type $type
     * @return \Zheltikov\TypeAssert\Parser\Node|null
     */
    public function &getFirstByType(Type $type): ?Node
    {
        /** @var \Zheltikov\TypeAssert\Parser\Node $child */
        foreach ($this->children as &$child) {
            if ($child->getType()->equals($type)) {
                return $child;
            }

            if ($child->hasChildren()) {
                $result =& $child->getFirstByType($type);
                if ($result !== null) {
                    return $result;
                }
            }
        }

        $result = null;
        return $result;
    }

    /**
     * @return \Zheltikov\TypeAssert\Parser\Type
     */
    public function getType(): Type
    {
        return $this->type;
    }

    /**
     * @param \Zheltikov\TypeAssert\Parser\Type $type
     * @return bool
     */
    public function hasChildOfType(Type $type): bool
    {
        foreach ($this->children as $child) {
            if ($child->getType()->equals($type)) {
                return true;
            }
        }

        return false;
    }

    /**
     * @param \Zheltikov\TypeAssert\Parser\Node $child
     * @return $this
     */
    public function deleteChild(Node $child): self
    {
        $new_children = [];

        foreach ($this->children as $_child) {
            if ($child !== $_child) {
                $new_children[] = $_child;
            }
        }

        $this->children = $new_children;

        return $this;
    }

    /**
     * @return bool
     */
    public function hasChildren(): bool
    {
        return $this->countChildren() > 0;
    }

    /**
     * @return int
     */
    public function countChildren(): int
    {
        return count($this->children);
    }

    /**
     * @return string
     */
    public function pretty(): string
    {
        $result = $this->type->getValue();
        $child_count = $this->countChildren();

        if ($child_count > 0) {
            $result .= '<';

            foreach ($this->children as $index => $child) {
                $result .= $child->pretty();

                if ($index + 1 < $child_count) {
                    $result .= ', ';
                }
            }

            $result .= '>';
        }

        return $result;
    }
}
